### PR TITLE
Validate timeColumnName when adding/updating schema/tableConfig

### PR DESCRIPTION
**Pull request by npawar**
## Description
https://github.com/apache/incubator-pinot/issues/5915
Add validation to check that timeColumnName is in sync across the table config and schema.

**Cases**
- Adding/updating tableConfig - Find associated schema (using rawTableName or from validationConfig.getSchemaName). Check that timeColumnName is present in schema. For OFFLINE, it is possible that timeColumnName/schema is null.
- Adding/updating schema - Find associated table configs (schemaName_OFFLINE and schemaName_REALTIME) Check timeColumnName used in all associated tableConfigs also exists in schema.

**Corner case** 
If tableConfig has been created with a schemaName != rawTableName, then we will not be able to validate during schema add/upload. 

## Release Notes
After this change, 
1. uploading a tableConfig with timeColumnName not present in schema will be disallowed (provided schema already uploaded)
2. uploading schema with timeColumnName not present in the schema will be disallowed (provided schemaName = tableName)

### CODE REVIEW

**Comment by mcvsubbu** [comment]
We do allow to have tables without a schema. I have not had the time to go through this fully, but do you account for the fact that a (offline) table may not have a schema at all?

**Comment by npawar** [comment]
> We do allow to have tables without a schema. I have not had the time to go through this fully, but do you account for the fact that a (offline) table may not have a schema at all?

Yes, I have accounted for offline table not having schema at all. I've put in the description all the cases it covers, and the case we miss out

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotSchemaRestletResource.java [hunk 1]
```diff
@@ -171,7 +173,8 @@ public SuccessResponse addSchema(
   public String validateSchema(FormDataMultiPart multiPart) {
     Schema schema = getSchemaFromMultiPart(multiPart);
     try {
-      SchemaUtils.validate(schema);
+      List<TableConfig> tableConfigs = getTableConfigsForSchema(schema.getSchemaName());
```
Comment (icefury71) [inline pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotSchemaRestletResource.java:176]: I wonder if this (along with the getTableConfigsForSchema function) can be moved inside validate ? So that the caller doesn't have to do this every single time (and can be done implicitly)

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTableIndexingConfigs.java [hunk 2]
```diff
@@ -58,7 +59,8 @@ public SuccessResponse updateIndexingConfig(
     TableConfig tableConfig;
     try {
       tableConfig = JsonUtils.stringToObject(tableConfigString, TableConfig.class);
-      TableConfigUtils.validate(tableConfig);
+      Schema schema = pinotHelixResourceManager.getSchemaForTableConfig(tableConfig);
```
Comment (icefury71) [inline pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTableIndexingConfigs.java:62]: Same here

File: pinot-core/src/main/java/org/apache/pinot/core/util/TableConfigUtils.java [hunk 3]
```diff
@@ -74,19 +80,37 @@ public static void validateTableName(TableConfig tableConfig) {
...
+    // timeColumnName can be null in OFFLINE table
+    if (timeColumnName != null && schema != null) {
+      Preconditions.checkState(schema.getSpecForTimeColumn(timeColumnName) != null,
+          "Cannot find valid fieldSpec for timeColumn: %s from the table config, in the schema: %s", timeColumnName,
+          schema.getSchemaName());
+    }
+
+    String peerSegmentDownloadScheme = validationConfig.getPeerSegmentDownloadScheme();
+    if (peerSegmentDownloadScheme != null) {
+      if (!CommonConstants.HTTP_PROTOCOL.equalsIgnoreCase(peerSegmentDownloadScheme) && !CommonConstants.HTTPS_PROTOCOL
```
Comment (icefury71) [inline pinot-core/src/main/java/org/apache/pinot/core/util/TableConfigUtils.java:110]: Curious - how is this related to time column field validation ?

File: pinot-core/src/test/java/org/apache/pinot/core/util/SchemaUtilsTest.java [hunk 4]
```diff
@@ -35,6 +43,115 @@
...
 
+  private static final String TABLE_NAME = "testTable";
+  private static final String TIME_COLUMN = "timeColumn";
+
+  @Test
+  public void testCompatibilityWithTableConfig() {
+    // empty list
+    List<TableConfig> tableConfigs = new ArrayList<>();
+    Schema schema = new Schema.SchemaBuilder().setSchemaName(TABLE_NAME).build();
+    SchemaUtils.validate(schema, tableConfigs);
```
Comment (icefury71) [inline pinot-core/src/test/java/org/apache/pinot/core/util/SchemaUtilsTest.java:54]: nit: wrap with try-catch for future proofing ?

File: pinot-core/src/test/java/org/apache/pinot/core/util/SchemaUtilsTest.java [hunk 5]
```diff
@@ -35,6 +43,115 @@
...
+
+    // offline table
+    // null timeColumnName
+    TableConfig tableConfig = new TableConfigBuilder(TableType.OFFLINE).setTableName(TABLE_NAME).build();
+    SchemaUtils.validate(schema, Lists.newArrayList(tableConfig));
```
Comment (icefury71) [inline pinot-core/src/test/java/org/apache/pinot/core/util/SchemaUtilsTest.java:59]: Same here

**Review by icefury71 — APPROVED** [review]
Minor comments

File: pinot-core/src/main/java/org/apache/pinot/core/util/TableConfigUtils.java (hunk 3 above)
Comment (npawar) [reply pinot-core/src/main/java/org/apache/pinot/core/util/TableConfigUtils.java:110]: This part was already present in the validateValidationConfig method (line 84 in the before diff)

File: pinot-core/src/test/java/org/apache/pinot/core/util/SchemaUtilsTest.java (hunk 4 above)
Comment (npawar) [reply pinot-core/src/test/java/org/apache/pinot/core/util/SchemaUtilsTest.java:54]: do you mean a case where the schema becomes invalid, but we pass the test anyway because of not having a try-catch?
Just to be sure, I tried this out. The test fails if there's any exception in the should-pass cases.
Lmk if you were referring to something else.

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotSchemaRestletResource.java (hunk 1 above)
Comment (npawar) [reply pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotSchemaRestletResource.java:176]: i thought about that. I was just trying to avoid passing pinotHelixResourceManager to the utils methods. I can change it to that if you think that's better. I was going back and forth about it

Comment (npawar) [reply pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotSchemaRestletResource.java:176]: Actually, as I tried to do this, I realized that PinotHelixResourceManager is in pinot-controller, and the utils (which are in pinot-core) do not have access to it. We'd have to pass propertyStore, or change the design of SchemaUtils so we can inject and keep state.
Now I think it's better that the caller fetches what it needs. Practically,  PinotTableRestletResource and PinotSchemaRestletResource should be the only places adding/modifying table/schema.

Comment (icefury71) [reply pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotSchemaRestletResource.java:176]: Understood. Thanks for looking into this.

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotSchemaRestletResource.java [hunk 6]
```diff
@@ -318,4 +324,19 @@ private void deleteSchemaInternal(String schemaName) {
           Response.Status.INTERNAL_SERVER_ERROR);
     }
   }
+
+  private List<TableConfig> getTableConfigsForSchema(@Nullable String schemaName) {
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotSchemaRestletResource.java:328]: `schemaName` should never be null

Comment (Jackie-Jiang) [reply pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotSchemaRestletResource.java:328]: Wondering if we can move this method into `PinotHelixResourceManager`?

**Review by Jackie-Jiang — APPROVED** [review]
LGTM with minor comments

File: pinot-core/src/test/java/org/apache/pinot/core/util/SchemaUtilsTest.java (hunk 4 above)
Comment (icefury71) [reply pinot-core/src/test/java/org/apache/pinot/core/util/SchemaUtilsTest.java:54]: I meant catching exception and explicitly 'fail'ing. It's fine though - just a nit.

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotSchemaRestletResource.java (hunk 6 above)
Comment (npawar) [reply pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotSchemaRestletResource.java]: yes, done
